File: src/server/codegen/model_gen/boosted_tree_ensemble.py
"""
Copyright 2017-2024 SensiML Corporation

This file is part of SensiML™ Piccolo AI™.

SensiML Piccolo AI is free software: you can redistribute it and/or
modify it under the terms of the GNU Affero General Public License
as published by the Free Software Foundation, either version 3 of
the License, or (at your option) any later version.

SensiML Piccolo AI is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with SensiML Piccolo AI. If not, see <https://www.gnu.org/licenses/>.
"""

import codegen.model_gen.utils as utils
from pandas import DataFrame


def create_boosted_tree_classifier_arrays(model_index, models):
    from library.classifiers.boosted_tree_ensemble import gtree_t

    outputs = []

    def assign_values(model_index, name, index, values, value_type="uint8_t"):
        node_count = len(values)
        return "{5} model_{0}_{1}_boosted_tree_{2}[{4}] = {{{3}}};".format(
            model_index, name, index, str(values)[1:-1], node_count, value_type
        )

    def clean(values):
        return [int(x) if x >= 0 else 0 for x in values]

    def clean_float(values):
        return [float(x) for x in values]

    tree_groups = DataFrame(models).groupby("Tree")

    for index, tree_model in tree_groups:
        # models.append(gtree_t(group))
        # for index, tree_model in enumerate(models):
        model = gtree_t(tree_model)
        outputs.append(
            assign_values(model_index, "node_list", index, clean(model.node_list))
        )
        outputs.append(
            assign_values(
                model_index,
                "leafs",
                index,
                clean_float(model.leafs),
                value_type="float",
            )
        )
        outputs.append(
            assign_values(model_index, "threshold", index, clean(model.threshold))
        )
        outputs.append(
            assign_values(model_index, "features", index, clean(model.features))
        )

    return outputs


def create_boosted_tree_classifier_struct(model_index, models):
    outputs = []

    tree_groups = DataFrame(models).groupby("Tree")

    outputs.append(
        "boosted_tree_t boosted_tree_ensemble_models_{0}[{1}] = ".format(
            model_index, len(tree_groups)
        )
        + "{"
    )
    for index, _ in enumerate(tree_groups):
        outputs.append("\t{")
        outputs.append(
            "\t\t.node_list = model_{1}_node_list_boosted_tree_{0},".format(
                index, model_index
            )
        )
        outputs.append(
            "\t\t.leafs = model_{1}_leafs_boosted_tree_{0},".format(index, model_index)
        )
        outputs.append(
            "\t\t.threshold = model_{1}_threshold_boosted_tree_{0},".format(
                index, model_index
            )
        )
        outputs.append(
            "\t\t.features = model_{1}_features_boosted_tree_{0},".format(
                index, model_index
            )
        )
        outputs.append("\t},")
    outputs.append("};")

    return outputs


def create_classifier_structures(kb_models):
    """
    typedef struct tree{
        uint8_t *node_list;
        float *leafs;
        uint8_t *threshold;
        uint8_t *features;
    } tree_t;


    typedef struct tree_ensemble_classifier_rows{
        uint8_t number_of_trees;
        tree_t * tree_ensemble;
    } tree_ensemble_classifier_rows_t;

    """
    output = []

    iterations = 0
    for model in kb_models:
        if (
            model["classifier_config"].get("classifier", "PME")
            == "Boosted Tree Ensemble"
        ):
            output.extend(
                create_boosted_tree_classifier_arrays(iterations, model["model_arrays"])
            )
            output.extend(
                create_boosted_tree_classifier_struct(iterations, model["model_arrays"])
            )
            iterations += 1

    iterations = 0
    output.append(
        (
            "boosted_tree_ensemble_classifier_rows_t boosted_tree_ensemble_classifier_rows[{}] = ".format(
                utils.get_number_classifiers(kb_models, "Boosted Tree Ensemble")
            )
            + "{"
        )
    )

    for model in kb_models:
        if (
            model["classifier_config"].get("classifier", "PME")
            == "Boosted Tree Ensemble"
        ):
            tree_groups = DataFrame(model["model_arrays"]).groupby("Tree")
            output.append("\n\t{")
            output.append("\t\t.number_of_trees = {},".format(len(tree_groups)))
            output.append(
                "\t\t.boosted_tree_ensemble = boosted_tree_ensemble_models_{},".format(
                    iterations
                )
            )
            output.append("\t},")
            iterations += 1

    output.append("};")

    return output


def validate_model_parameters(data):
    cleaned_data = data

    return cleaned_data


def validate_model_configuration(data):
    return data


def create_max_tmp_parameters(kb_models):
    return []


def get_output_tensor_size(model):
    return len(model["class_map"])

File: src/embedded_ml_sdk/classifiers/boosted_tree_ensemble/boosted_tree_ensemble.c
/*
Copyright 2017-2024 SensiML Corporation

This file is part of SensiML™ Piccolo AI™.

SensiML Piccolo AI is free software: you can redistribute it and/or
modify it under the terms of the GNU Affero General Public License
as published by the Free Software Foundation, either version 3 of
the License, or (at your option) any later version.

SensiML Piccolo AI is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with SensiML Piccolo AI. If not, see <https://www.gnu.org/licenses/>.
*/




#include "boosted_tree_ensemble.h"

boosted_tree_ensemble_classifier_rows_t *GBTClassifierTable;

float boosted_tree_classification(boosted_tree_t *model, uint8_t *feature_vector)
{

    int32_t next_node = 0;
    uint8_t current_node = 0;
    while (1)
    {
        next_node = model->node_list[current_node];
        if (feature_vector[model->features[current_node]] > model->threshold[current_node])
        {
            next_node += 1;
        }

        if (model->node_list[next_node] == 0)
        {
            return model->leafs[model->features[next_node]];
        }
        current_node = next_node;
    }
}

uint8_t boosted_tree_ensemble_classification(boosted_tree_t *forest_ensemble, uint8_t number_of_trees, uint8_t *feature_vector)
{

    uint8_t y = 0;
    uint8_t max_count = 0;
    float margin = 0.f;

    for (int32_t i = 0; i < number_of_trees; i++)
    {
        margin += boosted_tree_classification(&forest_ensemble[i], feature_vector);
    }

    return margin <= 0.f ? 1 : 2;
}

uint8_t boosted_tree_ensemble_simple_submit(uint8_t classifier_id, feature_vector_t *feature_vector, model_results_t *model_results)
{

    uint8_t y = 0;
    uint8_t max_count = 0;
    float margin = 0.f;

    for (int32_t i = 0; i < GBTClassifierTable[classifier_id].number_of_trees; i++)
    {
        margin += boosted_tree_classification(&GBTClassifierTable[classifier_id].boosted_tree_ensemble[i], (uint8_t *)feature_vector->data);
    }

    model_results->result = margin <= 0.f ? 1.0 : 2.0;

    return margin <= 0.f ? 1 : 2;
}

/*
 * Inititalize the PME
 */
void boosted_tree_ensemble_init(boosted_tree_ensemble_classifier_rows_t *classifier_table, const uint8_t num_classifiers)
{
    GBTClassifierTable = classifier_table;
}
